go/saltpack/decrypt.go

```diff
@@ -92,14 +92,30 @@ func (ds *decryptStream) read(b []byte) (n int, err error) {
 	return n, nil
 }
 
-func (ds *decryptStream) readHeader() error {
-	var hdr EncryptionHeader
-	seqno, err := ds.mps.Read(&hdr)
+func (ds *decryptStream) readHeader(rawReader io.Reader) error {
+	// Parse the length of the header.
+	var headerLength int
+	seqno, err := ds.mps.Read(&headerLength)
+	if err != nil {
+		return ErrFailedToDecodeHeaderLength
+	}
+	// Read the header bytes.
+	headerBytes := make([]byte, headerLength)
+	_, err = io.ReadFull(rawReader, headerBytes)
+	if err != nil {
+		return ErrFailedToReadHeaderBytes
+	}
+	// Compute the header hash.
+	headerHash := sha512.Sum512(headerBytes)
+	ds.headerHash = headerHash[:]
+	// Parse the header bytes.
+	var header EncryptionHeader
+	err = decodeFromBytes(&header, headerBytes)
```

Comment:
My idea is to use `bytes.Buffer` as a stream, and then assert that the # of bytes remaining is 0 after the read. This might give false positives but never false negatives.